src/core.lisp

```diff
@@ -1,3 +1,90 @@
+         (x (v:v. (scene:camera-x-vector camera) o*->p*))
+         (y (v:v. (scene:camera-y-vector camera) o*->p*)))
+    (make-depth-point :point
+                      (v:vec2 (round (* (/ (1+ x) 2) (scene:camera-width camera)))
+                              (round (* (/ (1+ y) 2) (scene:camera-height camera))))
+                      :depth
+                      (v:v. O->p
+                            (v:vunit (scene:camera-direction camera))))))
+
+(defun trg-to-camera (camera trg)
+  (mapcar (lambda (point)
+            (project-point-to-camera camera point))
+          trg))
+
+(defun triangle-contains (triangle point)
+  (let ((a->p (v:v- point (depth-point-point (first triangle))))
+        (b->p (v:v- point (depth-point-point (second triangle))))
+        (c->p (v:v- point (depth-point-point (third triangle))))
+        (bc-cross (v:vrot2 (v:v- (depth-point-point (second triangle))
+                                 (depth-point-point (third triangle)))
+                           (/ pi 2)))
+        (ca-cross (v:vrot2 (v:v- (depth-point-point (third triangle))
+                                 (depth-point-point (first triangle)))
+                           (/ pi 2)))
+        (ab-cross (v:vrot2 (v:v- (depth-point-point (first triangle))
+                                 (depth-point-point (second triangle)))
+                           (/ pi 2))))
+    (and (< 0 (* (v:v. a->p bc-cross)
+                 (v:v. b->p ca-cross)))
+         (< 0 (* (v:v. b->p ca-cross)
+                 (v:v. c->p ab-cross))))))
+
+(defun calculate-depth (point2d triangle)
+  "Calculate depth of a point given depths of triangle vertices."
+  (let ((distance-a (v:v2norm (v:v- point2d
+                                    (depth-point-point (first triangle)))))
+        (distance-b (v:v2norm (v:v- point2d
+                                    (depth-point-point (second triangle)))))
+        (distance-c (v:v2norm (v:v- point2d
+                                    (depth-point-point (third triangle))))))
+    (/ (+ (* (depth-point-depth (first triangle)) distance-b distance-c)
+          (* (depth-point-depth (second triangle)) distance-c distance-a)
+          (* (depth-point-depth (third triangle)) distance-a distance-b))
+       (+ (* distance-a distance-b)
+          (* distance-b distance-c)
+          (* distance-c distance-a)))))
+
+(defun render (scene camera)
+  (let ((image (make-array (list (scene:camera-width camera) (scene:camera-height camera))
+                           :initial-element '(0 0 0)))
+        (depths (make-array (list (scene:camera-width camera) (scene:camera-height camera))
+                            :initial-element :infinity)))
+    (dolist (triangle (scene:scene-triangles scene))
+      (let* ((flat-triangle (trg-to-camera camera triangle))
+             (minimal-x (max 0 (reduce #'min triangle :key #'d-p-x)))
+             (maximal-x (min (scene:camera-width camera)
```

Comment:
Should probably be `(1- (scene:camera-width camera))` instead of just width? And same for the height below.